src/realm/sync/subscriptions.cpp

```diff
@@ -0,0 +1,450 @@
+                throw std::logic_error("subscription set must be in Bootstrapping to update state to error");
+            }
+            if (!error_str) {
+                throw std::logic_error("Must supply an error message when setting a subscription to the error state");
+            }
+
+            m_obj.set(m_mgr->m_sub_set_keys->state, static_cast<int64_t>(new_state));
+            m_obj.set(m_mgr->m_sub_set_keys->error_str, *error_str);
+            break;
+        case State::Bootstrapping:
+        case State::Pending:
+            if (error_str) {
+                throw std::logic_error(
+                    "Cannot supply an error message for a subscription set when state is not Error");
+            }
+            m_obj.set(m_mgr->m_sub_set_keys->state, static_cast<int64_t>(new_state));
+            break;
+        case State::Complete:
+            if (error_str) {
+                throw std::logic_error(
+                    "Cannot supply an error message for a subscription set when state is not Error");
+            }
+            m_obj.set(m_mgr->m_sub_set_keys->state, static_cast<int64_t>(new_state));
+            m_mgr->supercede_prior_to(m_tr, version());
+            break;
+    }
+}
+
+SubscriptionSet SubscriptionSet::make_mutable_copy() const
+{
+    auto new_tr = m_tr->duplicate();
+    if (!new_tr->promote_to_write()) {
+        throw std::runtime_error("could not promote flexible sync metadata transaction to writable");
+    }
+
+    auto sub_sets = new_tr->get_table(m_mgr->m_sub_set_keys->table);
+    auto new_pk = sub_sets->maximum_int(sub_sets->get_primary_key_column()) + 1;
+
+    SubscriptionSet new_set_obj(m_mgr, std::move(new_tr), sub_sets->create_object_with_primary_key(Mixed{new_pk}));
+    for (const auto& sub : *this) {
+        new_set_obj.insert_sub_impl(sub.created_at(), sub.updated_at(), sub.name(), sub.object_class_name(),
+                                    sub.query_string());
+    }
+
+    return new_set_obj;
+}
+
+void SubscriptionSet::commit()
+{
+    if (m_tr->get_transact_stage() != DB::transact_Writing) {
+        throw std::logic_error("SubscriptionSet is not in a commitable state");
+    }
+    if (state() == State::Uncommitted) {
+        update_state(State::Pending, util::none);
+    }
+    m_tr->commit_and_continue_as_read();
+}
+
+SubscriptionStore::SubscriptionStore(DBRef db)
+    : m_db(std::move(db))
+    , m_sub_set_keys(std::make_unique<SubscriptionSetKeys>())
+    , m_sub_keys(std::make_unique<SubscriptionKeys>())
+{
+    auto tr = m_db->start_read();
+
+    auto schema_metadata_key = tr->find_table(c_flx_metadata_table);
+    if (!schema_metadata_key) {
+        tr->promote_to_write();
+
+        auto schema_metadata = tr->add_table(c_flx_metadata_table);
+        auto version_col = schema_metadata->add_column(type_Int, c_flx_meta_schema_version_field);
+        schema_metadata->create_object().set(version_col, int64_t(1));
+
+        auto sub_sets_table =
```

Comment:
This boilerplate could be much more readable if you chose to implement it at the object-store level where the concept of a schema is available. In fact, you could consider adding it to the existing sync metadata Realm. See https://github.com/realm/realm-core/pull/5006/files#diff-bfe713108b234cb50bdba0e2380b085357af740969bcd6a5948389b8fdb0672bR74

Reply:
I thought about that, and decided to not do that because it felt like a layering violation and I thought as a general rule that information that the sync client itself needs to work with in the same realm as the data the sync client is syncing.

I'm not rejecting your suggestion tho, I'll think about how to make this better.

Reply:
Do you intend to add this new FLX metadata to the same Realm as the query result state itself, or to a separate metadata Realm?

Reply:
If it is stored in the same Realm as the rest of the attached state, then this seems like a fine place for it. I just noticed that the verbosity of constructing the tables at this level detracts from being able to easily discern the schema. Would be nice if we had a C++ SDK :)

Reply:
Into the same realm as the query result.

My thinking was that these types will need to be accessed and mutated from within `ClientImpl` and friends, and I don't want to create a dependency cycle where the sync client depends on object store which depends on the sync client.

I do wish we had a way of describing schema in core that was more declarative without needing all of Object Store.

Reply:
Won't this create schema violations at the SDK level if its stored in the same Realm that users will be accessing?

Reply:
It should be stored in an unsyncable section of the Realm.

Reply:
Wouldn't that still fail local checks? I believe we still do _most_ of the standard schema validation even for synced realms.

Reply:
Upon further research (and chatting with Jonathan), we magically do not consider a table part of the schema if the table name is not prefixed with `class_`. 🙃

Reply:
At last the class_ prefix is useful for something. Letting us store data in tables that are separate from the user's schema was the original purpose of that and then we never really did anything where that was the correct approach.